File: python/x402/mechanisms/tvm/default_assets.py
"""Default USD-pegged assets for TVM payment schemes.

Index 0 of each network list is the ``"$0.10"`` default.
"""

from __future__ import annotations

from typing import TypedDict

from .codecs.common import normalize_address
from .constants import (
    DEFAULT_DECIMALS,
    TVM_MAINNET,
    TVM_TESTNET,
    USDT_MAINNET_MINTER,
    USDT_TESTNET_MINTER,
)


class TvmDefaultAsset(TypedDict):
    """Default stablecoin entry for TVM exact scheme."""

    asset: str
    decimals: int
    symbol: str


DEFAULT_ASSETS: dict[str, list[TvmDefaultAsset]] = {
    TVM_MAINNET: [
        {"asset": USDT_MAINNET_MINTER, "decimals": DEFAULT_DECIMALS, "symbol": "USDT"},
    ],
    TVM_TESTNET: [
        {"asset": USDT_TESTNET_MINTER, "decimals": DEFAULT_DECIMALS, "symbol": "USDT"},
    ],
}


def get_default_asset(network: str, symbol: str | None = None) -> TvmDefaultAsset:
    """Look up a default asset by network and optional ticker.

    Args:
        network: CAIP-2 network.
        symbol: Ticker; omit for the network default.

    Returns:
        Matching entry.

    Raises:
        ValueError: If network or ticker is unknown.
    """
    assets = DEFAULT_ASSETS.get(network)
    if not assets:
        raise ValueError(f"No default asset configured for network {network}")
    if not symbol:
        return assets[0]
    normalized = symbol.upper()
    for entry in assets:
        if entry["symbol"].upper() == normalized:
            return entry
    raise ValueError(f"No {symbol} default asset configured for network {network}")


def find_default_asset(asset: str, network: str) -> TvmDefaultAsset | None:
    """Reverse lookup by asset id (via ``normalize_address``) and network.

    Args:
        asset: Jetton master from payment requirements.
        network: CAIP-2 network.

    Returns:
        Matching entry, or ``None``.
    """
    assets = DEFAULT_ASSETS.get(network)
    if not assets:
        return None
    normalized = normalize_address(asset)
    for entry in assets:
        if normalize_address(entry["asset"]) == normalized:
            return entry
    return None

File: typescript/packages/mechanisms/hedera/src/defaultAssets.ts
import type {
  DefaultAsset,
  DefaultAssetTable,
  FindDefaultAsset,
  GetDefaultAsset,
} from "@x402/core/types";
import {
  HEDERA_MAINNET_CAIP2,
  HEDERA_MAINNET_USDC,
  HEDERA_TESTNET_CAIP2,
  HEDERA_TESTNET_USDC,
  HEDERA_USDC_DECIMALS,
} from "./constants";

export type HederaDefaultAsset = DefaultAsset;

/** Default USD-pegged assets by CAIP-2 network; index 0 is the `"$0.10"` default. */
export const DEFAULT_ASSETS: DefaultAssetTable<HederaDefaultAsset> = {
  [HEDERA_MAINNET_CAIP2]: [
    { asset: HEDERA_MAINNET_USDC, decimals: HEDERA_USDC_DECIMALS, symbol: "USDC" },
  ],
  [HEDERA_TESTNET_CAIP2]: [
    { asset: HEDERA_TESTNET_USDC, decimals: HEDERA_USDC_DECIMALS, symbol: "USDC" },
  ],
};

/**
 * Look up a default asset by network and optional ticker.
 *
 * @param network - CAIP-2 network
 * @param symbol - Ticker; omit for the network default
 * @returns Matching entry
 * @throws If network or ticker is unknown
 */
export const getDefaultAsset: GetDefaultAsset<HederaDefaultAsset> = (network, symbol?) => {
  const assets = DEFAULT_ASSETS[network];
  if (!assets || assets.length === 0) {
    throw new Error(`No default asset configured for network ${network}`);
  }
  if (!symbol) {
    return assets[0];
  }
  const normalized = symbol.toUpperCase();
  const match = assets.find(entry => entry.symbol.toUpperCase() === normalized);
  if (!match) {
    throw new Error(`No ${symbol} default asset configured for network ${network}`);
  }
  return match;
};

/**
 * Reverse lookup by asset id and network.
 *
 * @param asset - HTS token id from payment requirements
 * @param network - CAIP-2 network
 * @returns Matching entry, or undefined
 */
export const findDefaultAsset: FindDefaultAsset<HederaDefaultAsset> = (asset, network) => {
  const assets = DEFAULT_ASSETS[network];
  if (!assets) {
    return undefined;
  }
  return assets.find(entry => entry.asset === asset);
};
